### PR TITLE
Remove kubectl run generators

### PR DESCRIPTION
**Pull request by soltysh**
**What type of PR is this?**
/kind api-change
/kind cleanup
/kind deprecation
/priority backlog
/sig cli

**What this PR does / why we need it**:
Over a year ago we've deprecated all the generators from under `kubectl run` command except for the one responsible for generating a pod (see https://github.com/kubernetes/kubernetes/pull/68132). 
This PR removes all the generators and deprecates all the flags towards the goal to make `kubectl run` simpler.

**Special notes for your reviewer**:
/assign @liggitt @seans3 @pwittrock 

**Does this PR introduce a user-facing change?**:
```release-note
Remove all the generators from kubectl run. It will now only create pods. Additionally, deprecates all the flags that are not relevant anymore.
```

/cc @kubernetes/sig-cli-api-reviews @kubernetes/sig-cli-pr-reviews

### CODE REVIEW

**Comment by soltysh** [comment]
/hold
for visibility

**Comment by soltysh** [comment]
/sig architecture
@kubernetes/sig-architecture-api-reviews since this is dropping a few of conformance tests

File: test/e2e/kubectl/kubectl.go [hunk 1]
```diff
@@ -1727,134 +1573,6 @@ metadata:
...
-			})
-			framework.ExpectNoError(err)
-		})
-
-		/*
-			Release : v1.9
-			Testname: Kubectl, run deployment
-			Description: Command ‘kubectl run’ MUST create a deployment, with --generator=deployment, when a image name is specified in the run command. After the run command there SHOULD be a deployment that should exist with one container running the specified image. Also there SHOULD be a Pod that is controlled by this deployment, with a container running the specified image.
-		*/
-		framework.ConformanceIt("should create a deployment from an image ", func() {
```
Comment (liggitt) [inline test/e2e/kubectl/kubectl.go]: I'm surprised we had a conformance test using deprecated function... that shouldn't really have happened

Comment (soltysh) [reply test/e2e/kubectl/kubectl.go]: yeah, my bad I didn't catch it earlier :disappointed:

**Comment by sftim** [comment]
Does this need an issue to remove / revise documentation about generators from the website?
(there might be one already)

**Comment by sftim** [comment]
Thinking about this, any website page that recommends using `kubectl run` will need to change. There might be some left.

**Comment by soltysh** [comment]
> Does this need an issue to remove / revise documentation about generators from the website?
> (there might be one already)

Yes, docs pr will follow.

**Comment by joefitzgerald** [comment]
Will there be an alternate imperative way to generate the yaml for a deployment?

**Comment by smarterclayton** [comment]
/approve

Since this has had 6 releases of deprecation, is clearly communicated via the SIG, and the release notes are proper.

**Comment by liggitt** [comment]
this looks to have turned the kubectl skew e2e job red: https://prow.k8s.io/job-history/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-new-master-gci-kubectl-skew?buildId=1222912140172595202

* last successful job: https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-new-master-gci-kubectl-skew/1222620466083008515
* first failed job: https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-new-master-gci-kubectl-skew/1222635818254864384

**Comment by soltysh** [comment]
@liggitt from what I've checked the logs this needs to go all the way back to 1.16, so I've opened https://github.com/kubernetes/kubernetes/pull/87997 for 1.16 and https://github.com/kubernetes/kubernetes/pull/87996 for 1.17 removing just the tests which were removed in this PR.

**Comment by vinsguru** [comment]
@soltysh 

Hi.. Are we planning to add other options like `--port`, `--replicas` etc as part of `create deployment` commands ?

**Comment by sftim** [comment]
@soltysh is there a docs PR against the _dev.1-19_ branch that connects to this?

**Comment by soltysh** [comment]
> @soltysh is there a docs PR against the _dev.1-19_ branch that connects to this?

It's in the works.

**Comment by soltysh** [comment]
> Hi.. Are we planning to add other options like `--port`, `--replicas` etc as part of `create deployment` commands ?

You're more than welcome to open a PR adding it.

**Comment by alexellis** [comment]
Now that `kubectl run` has effectively been deprecated for creating deployments, I'm working through lots of docs to try and convert previously-working examples to the new syntax.

I am running into a bit of a stumbling-block here as there doesn't seem to be an equivalent to:

```
kubectl run nginx-1 --image=nginx --port=80 --restart=Always
```

This is how far I got before hitting a bit of a wall:

```
space-mini:docs alex$ kubectl create deployment nginx-1 --image=nginx --port=80 --restart=Always
Error: unknown flag: --port
See 'kubectl create deployment --help' for usage.
space-mini:docs alex$ kubectl create deployment nginx-1 --image=nginx  --restart=Always
Error: unknown flag: --restart
See 'kubectl create deployment --help' for usage.
space-mini:docs alex$ 
```

The only alternative I can think of that makes sense is to have a YAML file which is honestly going to be much more repetitive and harder to maintain.

Does anyone have any suggestions on how to convert samples like this?

**Comment by alexellis** [comment]
Here's an example of what we are going to need to put pretty much everywhere `kubectl run` was used for deployments. It seems rather unfortunate that "improving" the UX has made Kubernetes harder to use for project maintainers.

First create a deployment for Nginx.

For Kubernetes 1.17 and lower:

```bash
kubectl run nginx-1 --image=nginx --port=80 --restart=Always
```

For 1.18 and higher:

```bash
export DEPLOYMENT=nginx-1

(cat<<EOF
apiVersion: apps/v1
kind: Deployment
metadata:
  name: $DEPLOYMENT
  labels:
    app: nginx
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:1.14.2
        ports:
        - containerPort: 80
EOF
) | kubectl apply -f -
```

**Comment by caruccio** [comment]
You guys are lucky this is not a Linus Torvalds' project.

**Comment by NinjaPanda91** [comment]
Can we revert this PR ? 
--port, --replicas flags and a bunch of flags are missing and I'm studying for CKA exam and i just found out it uses 1.18.

We can raise this back once I'm done with the exam.

**Comment by sftim** [comment]
To be clear, the deprecation occurred in https://github.com/kubernetes/kubernetes/pull/68132 based on SIG CLI's consensus-based decision making.
If you'd like to _enhance_ `kubectl create`, or to suggest enhancements, I feel that those suggestions are much more likely to be accepted.

**Comment by soltysh** [comment]
https://github.com/kubernetes/kubernetes/pull/91113 should solve the issue @alexellis hope this helps your case :smiley_cat:

**Comment by alexellis** [comment]
Nice to have port back, what about the other key flags?

**Comment by sftim** [comment]
To suggest an enhancement for `kubectl`, you can visit https://github.com/kubernetes/kubectl/issues/new (and I hope that people will make these suggestions)

**Comment by supirman** [comment]
> Can we revert this PR ?
> --port, --replicas flags and a bunch of flags are missing and I'm studying for CKA exam and i just found out it uses 1.18.
> 
> We can raise this back once I'm done with the exam.

I was taking some Qwiklabs labs, some of kubertes lab rely on `kubectl run`. It seem that kubectl in Google Cloud Shell got update recently and the lab not updated to this issue.

For the workaround, you can just use the old version where `kubectl run` still working in old ways.

**Comment by tm1810** [comment]
kubectl run for deployments is being missed, can we get the features back in create please, this will make a lot of difference

**Comment by armujahid** [comment]
@tm1810 you can check https://github.com/kubernetes/kubectl/issues/898

**Pull request merged**